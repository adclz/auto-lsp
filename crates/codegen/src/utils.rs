pub(crate) fn sanitize_string(string: &str) -> String {
    if let Some(v) = PUNCTUATION.get(string) {
        v.to_string()
    } else if let Some(v) = RUST_KEYWORDS.get(string) {
        v.to_string()
    } else {
        let mut result = String::new();
        for c in string.chars() {
            if c == '_' {
                result.push(c);
            } else if let Some(v) = PUNCTUATION.get(&c.to_string()) {
                result.push_str(v);
            } else {
                result.push(c);
            }
        }
        result.replace(" ", "_")
    }
}

pub(crate) fn sanitize_string_to_pascal(string: &str) -> String {
    if let Some(v) = PUNCTUATION.get(string) {
        v.to_string()
    } else if let Some(v) = RUST_KEYWORDS.get(string) {
        v.to_string()
    } else {
        let mut result = String::new();
        for c in string.chars() {
            if c == '_' {
                result.push(c);
            } else if let Some(v) = PUNCTUATION.get(&c.to_string()) {
                result.push_str(v);
            } else {
                result.push(c);
            }
        }
        result
            .replace(" ", "_")
            .split('_')
            .map(|word| {
                let mut chars = word.chars();
                match chars.next() {
                    Some(first) => format!(
                        "{}{}",
                        first.to_ascii_uppercase(),
                        chars.collect::<String>()
                    ),
                    None => String::new(),
                }
            })
            .collect()
    }
}

pub static PUNCTUATION: phf::Map<&'static str, &'static str> = phf::phf_map! {
    "{" => "LeftCurly",
    "}" => "RightCurly",
    "(" => "LeftParen",
    ")" => "RightParen",
    "[" => "LeftBracket",
    "]" => "RightBracket",
    "," => "Comma",
    ":" => "Colon",
    ";" => "Semicolon",
    "." => "Dot",
    "@" => "At",
    "!" => "Bang",
    "#" => "Hash",
    "$" => "Dollar",
    "%" => "Percent",
    "^" => "Caret",
    "&" => "Ampersand",
    "*" => "Star",
    "-" => "Minus",
    "_" => "Underscore",
    "+" => "Plus",
    "=" => "Equal",
    ">" => "Greater",
    "<" => "Less",
    "|" => "Pipe",
    "~" => "Tilde",
    "/" => "Slash",
    "\\" => "Backslash",
    "//" => "SlashSlash",
    "//=>" => "SlashSlashEqual",
    "//=" => "SlashSlashEqual",
    "/=" => "SlashEqual",
    "/>" => "SlashGreater",
    "/?" => "SlashQuestion",
    "/??" => "SlashNullish",
    "/*" => "SlashStar",
    "*/" => "StarSlash",
    "+++" => "PlusPlusPlus",
    "!!" => "BangBang",
    "!!=" => "BangBangEqual",
    "!!?" => "BangBangQuestion",
    "!!??" => "BangBangNullish",
    "!!???" => "BangBangNullishQuestion",
    "?" => "Question",
    "->" => "Arrow",
    "=>" => "FatArrow",
    "++" => "PlusPlus",
    "--" => "MinusMinus",
    "&&" => "And",
    "||" => "Or",
    "==" => "EqualEqual",
    "!=" => "NotEqual",
    ">=" => "GreaterEqual",
    "<=" => "LessEqual",
    "===" => "TripleEqual",
    "!==" => "NotTripleEqual",
    "<<" => "ShiftLeft",
    ">>" => "ShiftRight",
    ">>>" => "ShiftRightUnsigned",
    "+=" => "PlusEqual",
    "-=" => "MinusEqual",
    "*=" => "StarEqual",
    "%=" => "PercentEqual",
    "&=" => "AmpersandEqual",
    "|=" => "PipeEqual",
    "^=" => "CaretEqual",
    "&&=" => "AndEqual",
    "||=" => "OrEqual",
    "??=" => "NullishEqual",
    "??" => "Nullish",
    "???" => "NullishQuestion",
    "**" => "StarStar",
    "**=" => "StarStarEqual",
    "<>" => "LessGreater",
    "<=>" => "LessGreaterEqual",
};

pub static RUST_KEYWORDS: phf::Map<&'static str, &'static str> = phf::phf_map! {
    "abstract" => "Abstract",
    "as" => "As",
    "async" => "Async",
    "await" => "Await",
    "break" => "Break",
    "const" => "Const",
    "continue" => "Continue",
    "crate" => "Crate",
    "dyn" => "Dyn",
    "else" => "Else",
    "enum" => "Enum",
    "extern" => "Extern",
    "false" => "False",
    "final" => "Final",
    "fn" => "Fn",
    "for" => "For",
    "if" => "If",
    "impl" => "Impl",
    "in" => "In",
    "let" => "Let",
    "loop" => "Loop",
    "match" => "Match",
    "mod" => "Mod",
    "move" => "Move",
    "mut" => "Mut",
    "pub" => "Pub",
    "ref" => "Ref",
    "return" => "Return",
    "self" => "Self",
    "static" => "Static",
    "struct" => "Struct",
    "super" => "Super",
    "trait" => "Trait",
    "true" => "True",
    "type" => "Type",
    "unsafe" => "Unsafe",
    "use" => "Use",
    "where" => "Where",
    "while" => "While",
    "with" => "With",
    "yield" => "Yield",
    "None" => "_None",
    "Some" => "_Some",
    "Ok" => "_Ok",
    "Err" => "_Err",
    "Result" => "_Result",
    "Option" => "_Option",
    "Vec" => "_Vec",
};
