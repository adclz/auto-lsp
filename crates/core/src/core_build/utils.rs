/*
This file is part of auto-lsp.
Copyright (C) 2025 CLAUZEL Adrien

auto-lsp is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>
*/

pub(crate) fn intersecting_ranges(
    range1: &std::ops::Range<usize>,
    range2: &tree_sitter::Range,
) -> bool {
    range1.start <= range2.start_byte && range1.end >= range2.end_byte
}

pub(crate) fn tree_sitter_range_to_lsp_range(range: &tree_sitter::Range) -> lsp_types::Range {
    let start = range.start_point;
    let end = range.end_point;
    lsp_types::Range {
        start: lsp_types::Position {
            line: start.row as u32,
            character: start.column as u32,
        },
        end: lsp_types::Position {
            line: end.row as u32,
            character: end.column as u32,
        },
    }
}
