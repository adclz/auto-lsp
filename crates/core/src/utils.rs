#[macro_export]
macro_rules! dispatch_once {
    ($node:expr, [$($ty:ty => $method:ident($($param:expr),*)),*]) => {
        $(
            if let Some(n) = $node.downcast_ref::<$ty>() {
                return n.$method($($param),*);
            };
        )*
    };
}

#[macro_export]
macro_rules! dispatch {
    ($node:expr, [$($ty:ty => $method:ident($($param:expr),*)),*]) => {
        $(
            if let Some(n) = $node.downcast_ref::<$ty>() {
                n.$method($($param),*)?;
            };
        )*
    };
}