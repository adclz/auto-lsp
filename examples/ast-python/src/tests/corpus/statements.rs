/*
This file is part of auto-lsp.
Copyright (C) 2025 CLAUZEL Adrien

auto-lsp is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>
*/

use super::utils::Result;
use crate::snap;
use rstest::rstest;

#[rstest]
#[case("import a, b")]
#[case("import b.c as d")]
#[case("import a.b.c")]
fn import_statements(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("from a import b")]
#[case("from a import *")]
#[case("from a import (b, c)")]
#[case("from a.b import c")]
#[case("from . import b")]
#[case("from .. import b")]
#[case("from .a import b")]
#[case("from ..a import b")]
fn import_from_statements(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("from __future__ import print_statement")]
#[case("from __future__ import python4")]
#[case("from __future__ import (absolute_import, division, print_function, unicode_literals)")]
fn import_future_statements(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("print a")]
#[case("print b, c")]
#[case("print 0 or 1, 1 or 0,")]
#[case("print 0 or 1")]
#[case("print not True")]
fn print_statements(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("print >> a")]
#[case(r#"print >> a, "b", "c""#)]
fn print_statements_with_redirection(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("assert a")]
#[case("assert b, c")]
fn assert_statements(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("a")]
#[case("b + c")]
#[case("1, 2, 3")]
#[case("1, 2, 3,")]
fn expression_statements(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("del a[1], b[2]")]
fn delete_statements(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("while true:\n  pass\n  break\n  continue")]
fn control_flow_statements(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("return a")]
#[case("return a + b, c")]
#[case("return not b")]
fn return_statements(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("if a:\n  b\n  c")]
fn if_statements(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("if a:\n  b\nelif c:\n  d\nelse:\n  f")]
#[case("if a:\n  b\nelse:\n  f")]
#[case("if a: b")]
#[case("if a: b; c")]
fn if_else_statements(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("if a:\n  if b:\n    c\n  else:\n    if e:\n      f")]
fn nested_if_statements(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("while a:\n  b\n  c")]
#[case("while c:\n  d\nelse:\n  e\n  f")]
fn while_statements(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("for line, i in lines:\n  print line\n  for character, j in line:\n    print character\nelse:\n  print x")]
#[case("for x, in [(1,), (2,), (3,)]:\n  x")]
fn for_statements(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("try:\n  a\nexcept b:\n  c\nexcept d as e:\n  f\nexcept g, h:\n  i\nexcept:\n  j")]
#[case("try:\n  a\nexcept b:\n  c\n  d\nelse:\n  e\nfinally:\n  f")]
#[case("try:\n  a\nexcept* b:\n  c\nexcept* d as e:\n  f\nelse:\n  g\nfinally:\n  h")]
fn try_statements(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("with a as b:\n  c")]
#[case("with (open('d') as d,\n      open('e') as e):\n  f")]
#[case("with e as f, g as h,:\n  i")]
fn with_statements(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("async def a():\n  b")]
#[case("async def c(d):\n  e")]
#[case("async def g(g, h,):\n  i")]
#[case("async def c(a: str):\n  a")]
#[case("async def c(a: b.c):\n  a")]
#[case("async def d(a: Sequence[T]) -> T:\n  a")]
#[case("async def i(a, b=c, *c, **d):\n  a")]
#[case("async def d(a: str) -> None:\n  return None")]
#[case("async def d(a:str=\"default\", b=c) -> None:\n  return None")]
fn async_functions(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("def e((a,b)):\n  return (a,b)")]
#[case("def e(*list: str):\n  pass")]
#[case("def e(**list: str):\n  pass")]
#[case("def f():\n  nonlocal a")]
#[case("def g(h, i, /, j, *, k=100, **kwarg):\n  return h,i,j,k,kwarg")]
#[case("def h(*a):\n  i((*a))\n  j(((*a)))")]
#[case("def foo():\n    pass \\\n\\\n\\")]
fn function_definitions(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("class A:\n  def b(self):\n    return c")]
#[case("class B():\n  pass")]
#[case("class B(method1):\n  def method1(self):\n    return")]
#[case("class C(method1, Sequence[T]):\n  pass")]
#[case("class D(Sequence[T, U]):\n  pass")]
fn class_definitions(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("class A(B, C):\n  def d():\n    e")]
fn class_definitions_with_superclasses(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("@a.b\nclass C:\n  @d(1)\n  @e[2].f.g\n  def f():\n    g")]
#[case(" @f()\n  async def f():\n    g")]
#[case("@buttons[0].clicked.connect\ndef spam():\n    ...")]
fn decorated_definitions(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("raise")]
#[case("raise RuntimeError('NO')")]
#[case("raise RunTimeError('NO') from e")]
fn raise_statements(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("global a")]
#[case("global a, b")]
fn global_statements(#[case] input: &str) -> Result {
    snap!(input)
}

#[rstest]
#[case("exec '1+1'")]
#[case("exec 'x+=1' in None")]
#[case("exec 'x+=1' in a, b")]
#[case("exec func in {}")]
fn exec_statements(#[case] input: &str) -> Result {
    snap!(input)
}
