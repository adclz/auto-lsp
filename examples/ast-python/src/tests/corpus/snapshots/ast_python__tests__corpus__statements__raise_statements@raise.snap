---
source: src/tests/corpus/statements.rs
expression: module
---
Module {
    children: [
        SimpleStatement(
            RaiseStatement(
                RaiseStatement {
                    cause: None,
                    children: None,
                    [RANGE]
                },
            ),
        ),
    ],
    [RANGE]
}
